core/components/addressable_pq.hpp

```diff
@@ -0,0 +1,243 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_COMPONENTS_ADDRESSABLE_PQ_HPP_
+#define GKO_CORE_COMPONENTS_ADDRESSABLE_PQ_HPP_
+
+
+#include <algorithm>
+#include <vector>
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+
+
+/**
+ * An addressable priority queue based on a k-ary heap.
+ *
+ * It allows inserting key-value pairs, modifying their key as well as accessing
+ * and removing the key-value pair with the minimum key.
+ *
+ * @tparam KeyType    The type of the keys
+ * @tparam ValueType  The type of the values, it needs to be an integer type.
+ * @tparam deg_log2  The binary logarithm of the node degree k
+ */
+template <typename KeyType, typename ValueType, int deg_log2 = 4>
+struct addressable_priority_queue {
+    constexpr static int degree = 1 << deg_log2;
```

Comment:
I would advise on adding a static assert for `deg_log2` (to avoid UB or any overflow):
```suggestion
template <typename KeyType, typename ValueType, int deg_log2 = 4>
struct addressable_priority_queue {
    static_assert(0 < deg_log2 && deg_log2 < 31, "deg_log2 must be in [1, 30].");
    constexpr static int degree = 1 << deg_log2;
```

Reply:
that's UB and not allowed in a constexpr expression, see https://godbolt.org/z/bc1T6Yf9G